File: AOJ/DSL/1_a.cpp
#include <bits/stdc++.h>
#define _GLIBCXX_DEBUG
#define rep(i, n) for(int i = 0; i < (n); ++i)
#define rep1(i, n) for(int i = 1; i < (n); ++i)
using namespace std;
typedef long long ll;

struct UnionFind{
    vector<int> p;
    vector<int> r;

    UnionFind(int n){
        p.resize(n, -1);
        r.resize(n, 1);
    }

    int find(int x){
        if(p[x] == -1) return x;
        else return p[x] = find(p[x]);
    }

    void unite(int x, int y){
        x = find(x);
        y = find(y);

        if(x == y) return;
        if(r[x] > r[y]) swap(x, y);
        if(r[x] == r[y]) ++r[y];

        p[x] = y;
    }

    int same(int x, int y){
        x = find(x);
        y = find(y);

        return (x == y);
    }
};

int main(){
    cin.tie(nullptr);
    ios::sync_with_stdio(false);

    int n, q;
    cin >> n >> q;

    UnionFind uf(n);
    rep(_, q){
        int com, x, y;
        cin >> com >> x >> y;

        if(com == 0){
            uf.unite(x, y);
        }

        else{
            cout << uf.same(x, y) << endl;
        }
    }

    return 0;
}

File: AOJ/DSL/1_a.py
import sys
sys.setrecursionlimit(10**6)
class UnionFind:
    
    def __init__(self, n: int) -> None:
        self.p = [-1] * n

    def find(self, x: int) -> int:
        if(self.p[x] == -1): return x
        self.p[x] = self.find(self.p[x])
        return self.p[x]
    
    def unite(self, x: int, y: int) -> None:
        x = self.find(x)
        y = self.find(y)
        if(x == y): return
        
        self.p[x] = y



def main():
    n, q = map(int, input().split())
    uf = UnionFind(n)

    for _ in range(q):
        com, x, y = map(int, input().split())

        if(com == 0):
            uf.unite(x, y)
        
        if(com == 1):
            if(uf.find(x) == uf.find(y)): print(1)
            else: print(0)

if(__name__ == '__main__'):
    main()